pageserver/src/tenant/timeline.rs

```diff
@@ -118,6 +120,90 @@ impl PartialOrd for Hole {
     }
 }
 
+pub struct LayerFileManager(HashMap<PersistentLayerKey, Arc<dyn PersistentLayer>>);
+
+impl LayerFileManager {
+    fn get_from_desc(&self, desc: &PersistentLayerDesc) -> Arc<dyn PersistentLayer> {
+        // The assumption for the `expect()` is that all code maintains the following invariant:
+        // A layer's descriptor is present in the LayerMap => the LayerFileManager contains a layer for the descriptor.
+        self.0.get(&desc.key()).expect("not found").clone()
+    }
+
```

Comment:
At a minimum, the expect message should contain as much information as possible to figure out what went wrong here. E.g. layer file name to grep the logs for it.

I'm still not super comfortable with the reliance on the programmer to maintain the invariant in the comment.
Can we somehow prevent insertion into the LayerMap if the LayerFileManager doesn't contain a corresponding LayerDesc?

Reply:
fixed in https://github.com/neondatabase/neon/pull/4437/commits/c5d2c4fcc101e4fc1ee80c2f8857726f0a6c699d. Also added a helper function to insert into file mgr + layer map at the same time. This function will be part of LayerMapManager and the layer map modifications will be hidden from user at that time.

Reply:
I like the helper function.

Regarding my

> At a minimum, the expect message should contain as much information as possible to figure out what went wrong here. E.g. layer file name to grep the logs for it.

I don't yet see us logging layer file name or similar in the `expect`.

Reply:
oh I forgot to add that...